cuqi/implicitprior/_restorator.py

```diff
@@ -0,0 +1,220 @@
+    Moreau-Yoshida smoothing.
+    
+    We consider a density such that - \log\pi(x) = -g(x) with g convex, lsc,
+    proper but not differentiable. Consequently, we cannot apply any
+    algorithm requiring the gradient of g.
+    Idea:
+    We consider the Moreau envelope of g defined as
+    
+    g_{smoothing_strength} (x) = inf_z 0.5*\| x-z \|_2^2/smoothing_strength + g(z).
+    
+    g_{smoothing_strength} has some nice properties
+        - g_{smoothing_strength}(x)-->g(x) as smoothing_strength-->0 for all x
+        - \nabla g_{smoothing_strength} is 1/smoothing_strength-Lipschitz
+        - \nabla g_{smoothing_strength}(x) = (x - prox_g^{smoothing_strength}(x))/smoothing_strength for all x with 
+        
+        prox_g^{smoothing_strength}(x) = argmin_z 0.5*\| x-z \|_2^2/smoothing_strength + g(z) .
+
+    Consequently, we can apply any gradient-based algorithm with
+    g_{smoothing_strength} in lieu of g. These algorithms do not require the
+    full knowledge of g_{smoothing_strength} but only its gradient. The gradient
+    of g_{smoothing_strength} is fully determined by prox_g^{smoothing_strength}
+    and smoothing_strength.
+    It is important as, although there exists an explicit formula for
+    g_{smoothing_strength}, it is rarely used in practice, as it would require
+    us to solve an optimization problem each time we want to 
+    estimate g_{smoothing_strength}. Furthermore, there exist cases where we dont't
+    the regularization g with which the mapping prox_g^{smoothing_strength} is
+    associated.
+
+    Remark (Proximal operators are denoisers):
+        We consider the denoising inverse problem x = u + n, with
+        n \sim \mathcal{N}(0, smoothing_strength I).
+        A mapping solving a denoising inverse problem is called denoiser. It takes
+        the noisy observation x as an input and returns a less noisy version of x
+        which is an estimate of u.
+        We assume a prior density \pi(u) \propto exp(- g(u)).
+        Then the MAP estimate is given by 
+            x_MAP = \argmin_z 0.5 \| x - z \|_2^2/smoothing_strength + g(z) = prox_g^smoothing_strength(x)
+        Then proximal operators are denoisers. 
+    
+    Remark (Denoisers are not necessarily proximal operators): Data-driven
+    denoisers are not necessarily proximal operators
+    (see https://arxiv.org/pdf/2201.13256)
+    
+    Parameters
+    ---------- 
+    prior : RestorationPrior
+        Prior of the RestorationPrior type. In order to stay within the MYULA
+        framework the restorator of RestorationPrior must be a proximal operator.
+        
+    smoothing_strength : float
+        Smoothing strength of the Moreau-Yoshida envelope of the prior potential.
+    """
+
+    def __init__(self, prior:RestorationPrior, smoothing_strength=0.1, 
+                 **kwargs):
+        self.prior = prior
+        self.smoothing_strength = smoothing_strength
+
+        # if kwargs does not contain the geometry,
```

Comment:
Something here is confusing me. Why do we need to set the geometry like that? Why not rely entirely on the geometry of the prior?

Reply:
Good catch, thanks! Now we entirely rely on the geometry of the prior.